### PR TITLE
App + Subscription Metadata Editing, and User creation on login

### PR DESCRIPTION
**Pull request by Charlesthebird**
This PR includes:
- An improvement to the login UX with the request for initial user creation happening from the UI.
- Sections for editing app and subscription metadata.

This video shows the new login flow. I log in as the `admin_user`. This user does not have a record in the portal server DB, so the website sends a request to create the user and mutates the SWR key for `/me`. The user info is then displayed correctly on the website.

https://github.com/user-attachments/assets/3a032b75-1c00-4316-94d2-9e7152d7acb5

This is a video of editing the rate limit and custom metadata as an admin user:

https://github.com/user-attachments/assets/d9e709c5-b9f0-43a6-a1d7-40f06b3682a6

Non admin users see this apps page and app details page view of the metadata.
Screenshot 2024-09-27 at 2 43 32 PM

Screenshot 2024-09-27 at 2 43 45 PM

There is a portal server bug with the subscriptions not returning the metadata, so when that is completed the metadata editor should work on the subscriptions as well, since it uses the same logic.

Screenshot 2024-09-27 at 3 00 57 PM
BOT NOTES: 
resolves https://github.com/solo-io/solo-projects/issues/6952
resolves https://github.com/solo-io/solo-projects/issues/6958

### CODE REVIEW

File: projects/ui/src/Apis/gg_hooks.ts [hunk 1]
```diff
@@ -506,3 +507,142 @@ export function useDeleteOAuthMutation(appId: string) {
...
+    if (arg.rateLimit !== undefined) {
+      req.rateLimit = arg.rateLimit;
+    }
+    await fetchJSON(`/subscriptions/${arg.subscription.id}/metadata`, {
+      method: "POST",
+      headers: getLatestAuthHeaders(latestAccessToken),
+      body: JSON.stringify(arg),
+    });
+    mutate(`/apps/${arg.subscription.applicationId}/subscriptions`);
+    mutate(`/subscriptions?status=${SubscriptionStatus.APPROVED}`);
```
Comment (Worfle) [inline projects/ui/src/Apis/gg_hooks.ts:613]: This feels weird, doing all 3 mutates at once.  Not a real issue, but could use a comment to explain such a strong action

File: projects/ui/src/Utility/AdminUtility/RateLimitEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    const newRateLimitInfo = {
+      requestsPerUnit: requestsPerUnit.toString(),
+      unit,
+    };
+    const newCustomMetadata = customMetadata ?? {};
+    // Stop here if the object wasn't edited.
+    if (shallowEquals(newRateLimitInfo, initialRateLimitInfo)) {
+      return;
+    }
+    (async () => {
```
Comment (Worfle) [inline projects/ui/src/Utility/AdminUtility/RateLimitEditor.tsx:65]: Small nit that this section has a LOT of repeat that could be done as like `toastObject(verb: string, metadataType:string)` or whatever

File: projects/ui/src/Utility/utility.ts [hunk 3]
```diff
@@ -167,3 +167,19 @@ export const useInArea = (paths: string[]) => {
     });
   }, [routerLocation.pathname, paths]);
 };
+
+export const shallowEquals = (
```
Comment (Worfle) [inline projects/ui/src/Utility/utility.ts:171]: Is this not out of the box in JS yet?

Comment (Worfle) [reply projects/ui/src/Utility/utility.ts:171]: It still isn't. Sigh

File: projects/ui/src/Utility/utility.ts [hunk 4]
```diff
@@ -167,3 +167,19 @@ export const useInArea = (paths: string[]) => {
...
+  b: Record<string, any> | undefined
+) => {
+  if (!a && !b) {
+    return true;
+  }
+  if (!a || !b) {
+    return false;
+  }
+  return (
+    Object.entries(a).every(([k, v]) => b[k] === v) &&
```
Comment (Worfle) [inline projects/ui/src/Utility/utility.ts:182]: This feels needlessly duplicated.  It's slight, but terser would just be to compare the lengths above, like. for the `(!a || !b)` to do:  `(a?.length !== b?.length)`, and then when we get to this check then they must be the same length and we're only doing 1 iteration

File: projects/ui/src/Utility/AdminUtility/RateLimitEditor.tsx (hunk 2 above)
Comment (Charlesthebird) [reply projects/ui/src/Utility/AdminUtility/RateLimitEditor.tsx:65]: Agreed - I can do that refactor in a follow up.

**Pull request merged**